Fix compute persist_sink shutdown

teskje opened the pull request:
This PR makes persist_sink shut itself done correctly, both when the shutdown token is dropped, and when the inputs have advanced to the empty frontier.

The PR is split into 5 commits:

* The first commit makes a couple minor cleanups to the persist_sink. Those are not required but make the later changes slightly simpler.
* The second commit makes the persist_sink use the "state-of-the-art" approach to async timely operators. See [this Slack thread](https://materializeinc.slack.com/archives/CMNT2QSBZ/p1669653900369139) for context. The main difference is that we get rid of the `scheduler` and instead can write code that looks like a regular (non-timely) async function.
* The third commit adds logic to shut down the persist sink operators once all their inputs have stopped producing data.
* The fourth commit removes the persist writer heartbeating, which was broken by the second commit and is not needed anymore.
* The fifth commit adds a test that confirms dataflow cleanup is working as intended.

Apart from the dataflow operator shutdown, I also verified that with these changes there are no stray tokio tasks left behind either. So this is a potential fix for the seqno leak we are still seeing in production.

### Example

Materialized view dataflows get cleaned up as soon as the inputs have advanced to the empty frontier:

```sql
materialize=> create materialized view test as select generate_series(1, 1000);
CREATE MATERIALIZED VIEW
materialize=> select * from mz_internal.mz_dataflow_operators;
 id | worker_id | name
----+-----------+------
(0 rows)

materialize=> select count(*) from test;
 count
-------
  1000
(1 row)

materialize=> drop materialized view test;
DROP MATERIALIZED VIEW
```

Materialized view dataflows with inputs that don't advance to the empty frontier get cleaned up when they are dropped:

```sql
materialize=> create materialized view test as select * from mz_views;
CREATE MATERIALIZED VIEW
materialize=> select * from mz_internal.mz_dataflow_operators;
 id  | worker_id |                     name
-----+-----------+----------------------------------------------
 326 | 0         | Map
 328 | 0         | Map
 340 | 0         | Map
 342 | 0         | Map
 358 | 0         | Map
 322 | 0         | OkErr
 338 | 0         | OkErr
 354 | 0         | FlatMap
 347 | 0         | Feedback
 356 | 0         | Exchange
 330 | 0         | Concatenate
 344 | 0         | Concatenate
 324 | 0         | InspectBatch
 351 | 0         | InspectBatch
 363 | 0         | InspectBatch
 316 | 0         | Dataflow: 1.3.test
 370 | 0         | Dataflow: 1.3.test
 353 | 0         | persist_sink u2 write_batches
 365 | 0         | persist_sink u2 append_batches
 334 | 0         | persist_source u2: part fetcher
 318 | 0         | persist_source s275: part fetcher
 333 | 0         | persist_source u2: part distribution
 317 | 0         | persist_source s275: part distribution
 348 | 0         | persist_sink u2 mint_batch_descriptions
 336 | 0         | persist_source u2: consumed part collector
 320 | 0         | persist_source s275: consumed part collector

materialize=> drop materialized view test;
DROP MATERIALIZED VIEW
materialize=> select * from mz_internal.mz_dataflow_operators;
 id | worker_id | name
----+-----------+------
(0 rows)
```

### Motivation

  * This PR fixes a recognized bug.

Fixes https://github.com/MaterializeInc/materialize/issues/16326.
May help with https://github.com/MaterializeInc/materialize/issues/15937.

### Tips for reviewer

Look at the commits and their messages separately!

For the second commit, I suggest you view the diff with "Hide whitespace" enabled. That makes it less painful, but still not great unfortunately.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - <!-- Add release notes here or explicitly state that there are no user-facing behavior changes. -->

antiguru reviewed (approved):
This looks great!

lluki commented:
This cleans up also all operators if the source produces an empty frontier. Commands I ran (cc @vmarcos):
```
CREATE SOURCE lgtpch FROM LOAD GENERATOR TPCH (SCALE FACTOR 0.01) FOR ALL TABLES WITH (SIZE = '1');

CREATE MATERIALIZED VIEW q14 AS
SELECT
    100.00 * sum(case
        when p_type like 'PROMO%'
            then l_extendedprice * (1 - l_discount)
        else 0
    end) / sum(l_extendedprice * (1 - l_discount)) AS promo_revenue
FROM
    lineitem,
    part
WHERE
    l_partkey = p_partkey
    AND l_shipdate >= DATE '1995-09-01'
    AND l_shipdate < DATE '1995-09-01' + INTERVAL '1' month;

# Wait until the materialized view has produced data.
SELECT COUNT(*) FROM q14; 

# At this point there should be no more operators left.
SELECT * FROM mz_internal.mz_dataflow_operators;
```

this could probably be added like this to sqllogictest.

aljoscha commented:
looped in @petrosagg for the async operator usage

aljoscha commented on src/compute/src/sink/persist_sink.rs:
```diff
@@ -278,233 +274,239 @@ where
-                    current_desired_frontier.clone_from(desired_frontier);
-                    current_persist_frontier.replace(persist_frontier.clone());
-                } else {
-                    // WIP: Remove this!
```
:see_no_evil:

aljoscha commented on src/compute/src/sink/persist_sink.rs:
```diff
@@ -570,258 +568,313 @@ where
+                biased;
+                _ = shutdown_rx.as_mut() => return,
+
+                // Ensure that we are periodically woken up and keep heartbeating the persist
```
We could get rid of this because all handles now have a heartbeat task. But if it's not too much hassle we can leave it in because Dan has a long-term plan of removing those standard heartbeating tasks again.

teskje replied:
If it's a long-term plan, I'm for dropping them. Less code we have to maintain until that plan bears fruit!

teskje replied:
I removed the explicit heartbeating from the persist_sink now.

aljoscha replied:
The long term plan (at least for Dan) is to remove the automatic heartbeating. Which would require the manual heartbeating that we do here to remain in place.

aljoscha commented on src/compute/src/sink/persist_sink.rs:
```diff
@@ -570,258 +568,313 @@ where
+
+                // Ensure that we are periodically woken up and keep heartbeating the persist
+                // writer even if there are no changes in our inputs (updates or frontier changes).
+                _ = ticker.tick() => continue,
```
Could move the `write.maybe_heartbeat_writer().await;` to here.

teskje replied:
I was not sure about this. The comment above states that "We don't write to _Consensus_ but instead only write batch data to _Blob_ so someone might eventually expire our lease." I take that to mean that even if we see progress on the inputs, we still should heartbeat the writer? If that is the case, then we can't move the heartbeating here, because this branch is only taken if nothing else happens for longer than 60 seconds, which might never be the case for sufficiently busy inputs.

teskje replied:
Doesn't matter now that I have removed the heartbeating anyway!

aljoscha reviewed (commented):
Did a first pass and had some comments. Overall, using the new async operator API seems a bit clumsy/boilterplate-y here. I'm sure @petrosagg has some recommendations!

teskje commented:
I added a test for the dataflow cleanup now. This PR is ready for review!

petrosagg reviewed (commented):
This PR is almost correct but it relies on a subtle condition for it to actually shutdown the dataflows. When an operator is tokened and at the same time has inputs it is not sufficient to simply exit the async function to signal termination. This isn't specific to async operators, the equivalent in sync operators is that it is not sufficient to simply return `false` from the closure passed to `OperatorBuilder::build_reschedule` to signal termination.

In both cases timely will keep the operators alive as long as there are data at the input handles that have not been consumed. The reason for this is that the data in their inputs have associated input capabilities which overall have the effect of holding back the frontiers so it never reaches the empty antichain.

So for these kind of operators the response on a token being dropped is to enter a loop where inputs are continuously drained until they reach the empty antichain and only then exit the future.

I put up an example that showcases this problem here https://github.com/petrosagg/timely-dataflow/blob/cleanup-example/timely/examples/dataflow-cleanup.rs

If you toggle the comment that drains the input you can see the dataflow hanging or being properly cleaned up

teskje commented:
Rebased onto https://github.com/MaterializeInc/materialize/pull/16414 and changed token handling after [a discussion in Slack](https://materializeinc.slack.com/archives/CMNT2QSBZ/p1669894397794969).

teskje commented:
Rebased again onto main and removed the explicit heartbeating, as also discussed in above Slack thread.

petrosagg commented on src/compute/src/sink/persist_sink.rs:
```diff
@@ -881,234 +851,232 @@ where
+        if !active_worker {
+            // SUBTLE: This is different from `mint_batch_description`
+            // where the non-active workers have to stay alive and keep
+            // pumping input, to ensure that the one active worker sees
```
This comment is obsolete now. Nothing needs to stay alive to pump input as this is happening automatically so we can remove this comment

teskje replied:
Ah, that means we can also make `mint_batch_description` return early on inactive workers! I'll make that part of a separate PR, this one is already large enough.

petrosagg reviewed (approved):
So nice! Thank you

aljoscha commented:
Once Petros thinks this is good to merge I say we can merge! I haven't looked at the latest changes but what I gathered from offline discussions is enough for me.
